README
First time build:

autoreconf -i
automake -a
libtoolize
autoreconf


ChangeLog
Sep 4, 2010
0.1.0	Jim George <jgeorge@engr.colostate.edu>
* Added return values with specific error conditions
* Header file properly indicates packed data structures
* Init function does not attempt to init serial port now, expects an
already-configured serial port file descriptor



tests/test-amc.c
#include <stdlib.h>
#include <stdio.h>
#include <getopt.h>
#include <unistd.h>
#include <sys/ioctl.h>

#ifdef MOXA
#include <moxadevice.h>
#endif

#include "src/amc.h"

#ifdef MOXA
#warning Using MOXA-specific ioctls to set RS422 mode
#endif

char serial_device[256] = "/dev/ttyM0";
int baudrate = 115200;

int open_drive(struct amc_drive *drv, char *serial_device, int baudrate, int serial_mode, int *serial_fd)
{
	*serial_fd = amc_serial_open(serial_device, baudrate);

	if (*serial_fd == -1) {
		return -1;
	}

#ifdef MOXA
	ioctl(*serial_fd, MOXA_SET_OP_MODE, &serial_mode);
#endif

	if (0 != amc_drive_new(drv, 0x3F, *serial_fd)) {
		return -1;
	}

	amc_get_access_control(drv);

	return 0;
}

#define KP 30.0
#define KI 1.0
#define KS 20000.0

#define COUNTS_PER_REV 4096.0

#define SCALE_DC1 (8192.0/KP)
#define SCALE_DS1 (131072.0/(KI * KS))

enum {
	OPT_GETID = 256,
	OPT_DEBUG,
	OPT_PORT,
	OPT_ENABLEBRIDGE,
	OPT_QUICKSTOP,
	OPT_RESETEVENTS,
	OPT_BRIDGESTATUS,
	OPT_GETIFACE,
	OPT_SETIFACE,
	OPT_GETMOTORSTATUS,
	OPT_SETSPEED,
	OPT_REG16,
	OPT_REG32,
};

char *usage_string = 
"Retrieve/control modbus registers on AMC servo drives\n"
"Usage:\n"
"--getid: Retrieve drive ID string and version numbers\n"
"--port=<dev>: Set serial port device to dev\n"
"--debug: Show serial comms debug messages\n"
"--bridgestatus: Retrieve power bridge status\n"
"--enablebridge[=n]: Enable the power bridge, n=0 disables, n=1 enables\n"
"--quickstop[=n]: Perform quick stop, n=0 disables, n=1 enables\n"
"--resetevents: Reset latched events, if any\n"
"--getinterfaceinput=<n>: Retrieve value at interface input n\n"
"--setinterfaceinput=<n,val>: Set value at interface input n to specified value\n"
"--getmotorstatus: Get the motor status\n"
"--setspeed=<n>: Set motor speed in rpm\n"
"--reg16=<reg[,val]>: Get or set a 16-bit register. reg is a 16-bit hex number\n"
"        If specified, val is a 16-bit hex number to write.\n"
"--reg32=<reg[,val]>: Get or set a 32-bit register. reg is a 16-bit hex number\n"
"        If specified, val is a 32-bit hex number to write.\n"
;

static struct option opt_lst[] = {
	{"debug", no_argument, 0, OPT_DEBUG},
	{"port", required_argument, 0, OPT_PORT},

	{"getid", no_argument, 0, OPT_GETID},
	{"bridgestatus", no_argument, 0, OPT_BRIDGESTATUS},
	{"enablebridge", optional_argument, 0, OPT_ENABLEBRIDGE},
	{"quickstop", optional_argument, 0, OPT_QUICKSTOP},
	{"resetevents", no_argument, 0, OPT_RESETEVENTS},
	{"getinterfaceinput", required_argument, 0, OPT_GETIFACE},
	{"setinterfaceinput", required_argument, 0, OPT_SETIFACE},
	{"getmotorstatus", no_argument, 0, OPT_GETMOTORSTATUS},
	{"setspeed", required_argument, 0, OPT_SETSPEED},
	{"reg16", required_argument, 0, OPT_REG16},
	{"reg32", required_argument, 0, OPT_REG32},

	{NULL, 0, 0, 0}
};

int main(int argc, char *argv[])
{
	struct amc_drive *drv;
	int opt_idx, opt_errors = 0, opt;
	int serial_mode = 0;
	int serial_fd;

#ifdef MOXA
	serial_mode = RS422_MODE;
#endif

	if (argc == 1) {
		puts(usage_string);
		return -1;
	}
	
	drv = malloc(sizeof(struct amc_drive));

	if (-1 == open_drive(drv, serial_device, baudrate, serial_mode, &serial_fd)) {
		printf("Could not open %s\n", serial_device);
		return -1;
	}

	while (-1 != (opt = getopt_long(argc, argv, "dv", opt_lst, &opt_idx))) {
		switch(opt) {
		case OPT_PORT:
			close(serial_fd);
			if (-1 == open_drive(drv, serial_device, baudrate, serial_mode, &serial_fd)) {
					printf("Could not open %s\n", serial_device);
					return -1;
			}
			break;
		case OPT_GETID:
			{
			char buffer[256];
			amc_get_string(drv, 0x0B, 0x00, buffer, 256);
			printf("Drive name: %s\n", buffer);

			struct amc_product_info pi;
			if (0 == amc_get_product_info(drv, &pi)) {
				printf("Control Board Name: %s [%s]\n", pi.control_board_name, pi.control_board_version);
				printf("Product Part Number: %s [%s]\n", pi.product_part_number, pi.product_version);
			}
			else {
				printf("Could not retrieve product info\n");
				return -1;
			}
			}
			break;
		case OPT_DEBUG:
			drv->debug = 1;
			break;		
		case OPT_ENABLEBRIDGE:
			{
				uint16_t bridge_status;
				if (0 > amc_get_uint16(drv, 0x01, 0x00, &bridge_status)) {
					printf("Could not read bridge status\n");
					return -1;
				}
				if ((optarg == NULL) || strtol(optarg, NULL, 10)) {
					bridge_status &= ~AMC_BC_INHIBIT;
				}
				else {
					bridge_status |= AMC_BC_INHIBIT;
				}
				if (0 > amc_write_uint16(drv, 0x01, 0x00, bridge_status)) {
					printf("Could not write bridge status\n");
					return -1;
				}
			}
			break;
		case OPT_QUICKSTOP:
			{
				uint16_t bridge_status;
				if (0 > amc_get_uint16(drv, 0x01, 0x00, &bridge_status)) {
					printf("Could not read bridge status\n");
					return -1;
				}
				if ((optarg == NULL) || strtol(optarg, NULL, 10)) {
					bridge_status |= AMC_BC_QUICKSTOP;
				}
				else {
					bridge_status &= ~AMC_BC_QUICKSTOP;
				}
				if (0 > amc_write_uint16(drv, 0x01, 0x00, bridge_status)) {
					printf("Could not write bridge status\n");
					return -1;
				}
			}
			break;
		case OPT_RESETEVENTS:
			{
				uint16_t bridge_status;
				if (0 > amc_get_uint16(drv, 0x01, 0x00, &bridge_status)) {
					printf("Could not read bridge status\n");
					return -1;
				}
				bridge_status |= AMC_BC_RESETEVENTS;
				if (0 > amc_write_uint16(drv, 0x01, 0x00, bridge_status)) {
					printf("Could not write bridge status\n");
					return -1;
				}
				bridge_status &= ~AMC_BC_RESETEVENTS;
				if (0 > amc_write_uint16(drv, 0x01, 0x00, bridge_status)) {
					printf("Could not write bridge status\n");
					return -1;
				}
			}
			break;
		case OPT_BRIDGESTATUS:
			{
				uint16_t bridge_status;
				if (0 > amc_get_uint16(drv, 0x01, 0x00, &bridge_status)) {
					printf("Could not read bridge control\n");
					return -1;
				}
				printf("Bridge control: 0x%04X, Bridge: %s, Brake: %s, QuickStop: %s\n",
					bridge_status,
					(bridge_status & AMC_BC_INHIBIT) ? "Inhibited" : "Enabled",
					(bridge_status & AMC_BC_BRAKE) ? "Enabled" : "Disabled",
					(bridge_status & AMC_BC_QUICKSTOP) ? "Active" : "Inactive");

				if (0 > amc_get_uint16(drv, 0x02, 0x00, &bridge_status)) {
					printf("Could not read bridge status\n");
					return -1;
				}
				printf("Bridge status: 0x%04X \t[%c] Bridge Enabled\t[%c] DynBrake\n"
						"\t\t[%c] Shunt Reg Enabled\t[%c] Positive Stop\t[%c] Negative Stop\n"
						"\t\t[%c] PosTorqueInh\t[%c] NegTorqueInh\t[%c] Ext Brake\n",
					bridge_status,
					(bridge_status & AMC_BS_ENABLED) ? 'X' : ' ',
					(bridge_status & AMC_BS_DYNBRAKE) ? 'X' : ' ',
					(bridge_status & AMC_BS_SHUNT) ? 'X' : ' ',
					(bridge_status & AMC_BS_POSSTOP) ? 'X' : ' ',
					(bridge_status & AMC_BS_NEGSTOP) ? 'X' : ' ',
					(bridge_status & AMC_BS_POSTORQUEINH) ? 'X' : ' ',
					(bridge_status & AMC_BS_NEGTORQUEINH) ? 'X' : ' ',
					(bridge_status & AMC_BS_EXTBRAKE) ? 'X' : ' ');

				if (0 > amc_get_uint16(drv, 0x02, 0x01, &bridge_status)) {
					printf("Could not read drive protection status\n");
					return -1;
				}
				printf("Drive protection status: 0x%04X\t[%c] Reset\t[%c] Internal Error\t[%c] Short Circuit\n"
					"\t[%c] Overcurrent\t[%c] Undervoltage\t[%c] Overvoltage\t\t[%c] Overtemp\n",
					bridge_status,
					(bridge_status & AMC_PS_RESET) ? 'X' : ' ',
					(bridge_status & AMC_PS_INTERROR) ? 'X' : ' ',
					(bridge_status & AMC_PS_SHORTCKT) ? 'X' : ' ',
					(bridge_status & AMC_PS_OVERCURRENT) ? 'X' : ' ',
					(bridge_status & AMC_PS_UNDERVOLTAGE) ? 'X' : ' ',
					(bridge_status & AMC_PS_OVERVOLTAGE) ? 'X' : ' ',
					(bridge_status & AMC_PS_OVERTEMP) ? 'X' : ' ');

				if (0 > amc_get_uint16(drv, 0x02, 0x02, &bridge_status)) {
					printf("Could not read system protection status\n");
					return -1;
				}
				printf("System protection status: 0x%04X\t[%c] Param Restore Error\t[%c] Param Store Error\n"
					"\t[%c] Motor Overtemp\t[%c] Feedback Error\t[%c] Overspeed\t[%c] Comms Error\n",
					bridge_status,
					(bridge_status & AMC_SS_RESTOREERR) ? 'X' : ' ',
					(bridge_status & AMC_SS_STOREERR) ? 'X' : ' ',
					(bridge_status & AMC_SS_MOTOROVERTEMP) ? 'X' : ' ',
					(bridge_status & AMC_SS_FEEDBACKERROR) ? 'X' : ' ',
					(bridge_status & AMC_SS_OVERSPEED) ? 'X' : ' ',
					(bridge_status & AMC_SS_COMMERR) ? 'X' : ' ');

				if (0 > amc_get_uint16(drv, 0x02, 0x03, &bridge_status)) {
					printf("Could not read drive status 1\n");
					return -1;
				}				
				printf("Drive status 1: 0x%04X\t[%c] Log Missed\t[%c] Commanded Inhibit\t[%c] User Inhibit\n"
					"\t[%c] Pos Inhibit\t[%c] Neg Inhibit\t[%c] Current Limit\t[%c] Cont Current Limit\n"
					"\t[%c] Current Loop Sat\t[%c] Cmd Dyn Brk\t[%c] User Dyn Brk\t[%c] Shunt Reg\n",
					bridge_status,
					(bridge_status & AMC_DS_LOGMISSED) ? 'X' : ' ',
					(bridge_status & AMC_DS_CMDINHIBIT) ? 'X' : ' ',
					(bridge_status & AMC_DS_USERINHIBIT) ? 'X' : ' ',
					(bridge_status & AMC_DS_POSINH) ? 'X' : ' ',
					(bridge_status & AMC_DS_NEGINH) ? 'X' : ' ',
					(bridge_status & AMC_DS_CURRENTLIM) ? 'X' : ' ',
					(bridge_status & AMC_DS_CONTCURRENT) ? 'X' : ' ',
					(bridge_status & AMC_DS_CLSAT) ? 'X' : ' ',
					(bridge_status & AMC_DS_CMDDYNBRAKE) ? 'X' : ' ',
					(bridge_status & AMC_DS_USERDYNBRAKE) ? 'X' : ' ',
					(bridge_status & AMC_DS_SHUNTREG) ? 'X' : ' ');

				if (0 > amc_get_uint16(drv, 0x02, 0x04, &bridge_status)) {
					printf("Could not read drive status 1\n");
					return -1;
				}
				printf("Drive status 2: 0x%04X\t[%c] Zero Velocity\t[%c] At Command\t[%c] Vel Following Error\n"
					"\t[%c] Pos Velocity Limit\t[%c] Neg Velocity Limit\t[%c] Cmd Profiler\n",
					bridge_status,
					(bridge_status & AMC_DS_ZEROVEL) ? 'X' : ' ',
					(bridge_status & AMC_DS_ATCMD) ? 'X' : ' ',
					(bridge_status & AMC_DS_VELOCITYERR) ? 'X' : ' ',
					(bridge_status & AMC_DS_POSVELOCITYLIM) ? 'X' : ' ',
					(bridge_status & AMC_DS_NEGVELOCITYLIM) ? 'X' : ' ',
					(bridge_status & AMC_DS_CMDPROFILER) ? 'X' : ' ');
			}
			break;
		case OPT_SETIFACE:
			{
				uint32_t interface_number;
				uint32_t interface_value;
				char *next_ptr;
				
				interface_number = strtol(optarg, &next_ptr, 10);
				interface_value = strtol(next_ptr + 1, NULL, 10);
				
				if (interface_number > 15) {
					printf("Interface number %d > 15\n", interface_number);
					return -1;
				}
				if (0 > amc_write_uint32(drv, 0x45, interface_number, interface_value)) {
					printf("Could not write to interface %d\n", interface_number);
					return -1;
				}
			}
			break;
		case OPT_GETIFACE:
			{				
				uint32_t interface_number;
				uint32_t interface_value;
				
				interface_number = strtol(optarg, NULL, 10);
				
				if (interface_number > 15) {
					printf("Interface number %d > 15\n", interface_number);
					return -1;
				}
				if (0 > amc_get_uint32(drv, 0x45, interface_number, &interface_value)) {
					printf("Could not read number %d\n", interface_number);
					return -1;
				}
				printf("Interface %2d = 0x%08X (%d)\n", interface_number, interface_value, interface_value);
			}
			break;
		case OPT_GETMOTORSTATUS:
			{
				int16_t current_demand, current_measured;
				if (0 > amc_get_uint16(drv, 0x10, 0x02, &current_demand)) {
					printf("Could not read motor current\n");
					return -1;
				}
				if (0 > amc_get_uint16(drv, 0x10, 0x03, &current_measured)) {
					printf("Could not read motor current\n");
					return -1;
				}
				printf("Current demand: %.2f, measured: %.2f\n", current_demand/SCALE_DC1, current_measured/SCALE_DC1);
				int32_t speed_measured;
				amc_get_uint32(drv, 0x11, 0x02, &speed_measured);
				printf("Speed: %.2f rpm (%d)\n", (speed_measured / SCALE_DS1) / COUNTS_PER_REV * 60.0, speed_measured);
			}
			break;
		case OPT_SETSPEED:
			{
				int32_t speed;
				speed = (atoi(optarg) * COUNTS_PER_REV / 60.0) * SCALE_DS1;
				amc_write_uint32(drv, 0x45, 0, speed);
			}
			break;
		case OPT_REG16:
			{
				uint16_t reg_num;
				uint16_t reg_val;
				char *next_ptr;
				
				reg_num = strtol(optarg, &next_ptr, 16);
				if (*next_ptr != 0) {
					reg_val = strtol(next_ptr + 1, NULL, 16);
					amc_write_uint16(drv, reg_num >> 8, reg_num & 0xFF, reg_val);
				}
				amc_get_uint16(drv, reg_num >> 8, reg_num & 0xFF, &reg_val);
				printf("Register %02X:%02X = %04X (%5d)\n", reg_num >> 8, reg_num & 0xFF, reg_val, reg_val);
			}
			break;
		case OPT_REG32:
				{
				uint16_t reg_num;
				uint32_t reg_val;
				char *next_ptr;
				
				reg_num = strtol(optarg, &next_ptr, 16);
				if (*next_ptr != 0) {
					reg_val = strtol(next_ptr + 1, NULL, 16);
					amc_write_uint32(drv, reg_num >> 8, reg_num & 0xFF, reg_val);
				}
				amc_get_uint32(drv, reg_num >> 8, reg_num & 0xFF, &reg_val);
				printf("Register %02X:%02X = %08X (%10d)\n", reg_num >> 8, reg_num & 0xFF, reg_val, reg_val);
			}
		break;
		default:
			opt_errors++;
			break;
		}
	}

	if (opt_errors) {
		puts(usage_string);
	}

	return 0;
}



src/crc.c
/**
\file crc.c
\brief Module to compute CRCs according to AMC protocol manual
\author Jim George



*/

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include "crc.h"

typedef float(*crcfnptr)(unsigned short, unsigned short, unsigned short);

/**
\brief Create a CRC lookup table based on a specified polynomial up to 16 bits
\param data Input data for CRC calculation
\param genpoly Polynomial (up to 16 bits) to generate CRC
\param accum Previous data from which CRC was calculated
*/
static unsigned short crchware(uint16_t data, uint16_t genpoly, uint16_t accum)
{
	static int i;
	
	data <<= 8;
	
	for (i = 8; i > 0; i--) {
		if ((data ^ accum) & 0x8000) {
			accum = (accum << 1) ^ genpoly;
		}
		else {
			accum <<= 1;
		}
		data <<= 1;
	}
	return accum;
}

/**
\brief Compute a CRC lookup table
\param poly Polynomial (up to 16 bits) to create CRC table
\return Pointer to allocated memory block with CRC table or NULL on failure

This function computes a lookup table that's used by amc_crc_check_word to
compute CRCs. The lookup table is 256 16-bit words long, and is allocated by
this function. A NULL is returned on memory allocation failure. The block
must be freed later on using free().
*/
unsigned short *amc_crc_mktable(uint16_t poly)
{
	unsigned short *crctable;
	int i;
	
	crctable = (uint16_t *)malloc(256 * sizeof(uint16_t));
	if (crctable == NULL) return NULL;
	
	for (i = 0; i < 256; i++) {
		crctable[i] = crchware(i, poly, 0);
	}
	return crctable;
}

/**
\brief Check CRC of a single 16-bit word
\param data Input data to verify CRC
\param *accumulator Pointer to temporary storage where previous data was checked
\param *crc_table Pointer to lookup table used to speed up CRC computations
*/
void amc_crc_check_word(uint16_t data, uint16_t *accumulator, uint16_t *crc_table)
{
	*accumulator = (*accumulator << 8) ^ crc_table[(*accumulator >> 8) ^ data];
}



src/serial.c
/**
\file serial.c
\brief Module for serial port routines
\author Jim George
*/

#include <unistd.h>
#include <stdio.h>
#include <strings.h>
#include <assert.h>
#include <stdlib.h>
#include <fcntl.h>
#include <termios.h>
#include <sys/stat.h>
#include <linux/serial.h>
#include <errno.h>
#include "serial.h"

#define SERIAL_PORT_SPD_TBL_MAX 18
static struct {
	int baud;
	speed_t ident;
} serial_port_speed_table[SERIAL_PORT_SPD_TBL_MAX] = {
{50, B50},
{75, B75},
{110, B110},
{134, B134},
{150, B150},
{200, B200},
{300, B300},
{600, B600},
{1200, B1200},
{1800, B1800},
{2400, B2400},
{4800, B4800},
{9600, B9600},
{19200, B19200},
{38400, B38400},
{57600, B57600},
{115200, B115200},
{230400, B230400}
};

/**
\brief Get interface speed macro from integer speed
\param speed Input speed, in baud
\return One of the interface speed constants from termios library,
or -1 on error

Converts an integer baud rate into one of the termios speed
macros. If the baud rate specified is not found, an error is
returned.
*/
static speed_t serial_port_get_speed(const unsigned int speed)
{
	int ctr;
	
	for (ctr = 0; ctr < SERIAL_PORT_SPD_TBL_MAX; ctr++) {
		if (speed == serial_port_speed_table[ctr].baud) {
			return serial_port_speed_table[ctr].ident;
		}
	}
	return -1;
}

/**
\brief Initialize serial port
\param *device_name Unix device name for the serial port to open
\param speed Speed at which to open port
\param *port Pointer to the file descriptor for the serial port
\return 0 on success, -1 on error

The port is opened with 8N1 settings (8-bit, no parity, 1 stop bit)
*/
int serial_port_init(const char *device_name,
	unsigned int speed,
	int *port)
{
	struct termios term_st;
	
	assert(device_name != NULL);
	assert(port != NULL);
	
	*port = open(device_name, O_RDWR | O_NOCTTY);
	if (*port < 0) {
		return -1;
	}
	
	bzero(&term_st, sizeof(struct termios));

	if (tcgetattr(*port, &term_st)) {
		perror("tcgetattr");
		return -1;
	}

	/* Set interface speed */
	{
		int spd_macro = serial_port_get_speed(speed);
		if (spd_macro == -1) return -1;
		cfsetispeed(&term_st, spd_macro);
		cfsetospeed(&term_st, spd_macro);
	}
	/* Enable raw mode output */
	cfmakeraw(&term_st);
	term_st.c_oflag &= ~OPOST;
	
	/* Don't own the terminal (^C won't kill us), Enable rcv, Drop DTR on close */
	term_st.c_cflag |= (CREAD | HUPCL);
	
	/* Set 8-bit port size */
	term_st.c_cflag &= ~CSIZE; term_st.c_cflag |= CS8;

	/* No parity, 1 stop bit */
	term_st.c_cflag &= ~PARENB; term_st.c_cflag &= ~CSTOPB;
	
	/* Turn off flow control, RS-485 does not use it */
	term_st.c_cflag &= ~CRTSCTS;
	term_st.c_iflag &= ~(IXON | IXOFF | IXANY);
	
	/* Min. 1 character for read to unblock */
	term_st.c_cc[VMIN] = 1;
	/* Set timeout to 1 second */
	term_st.c_cc[VTIME] = 10;
	
	if (tcflush(*port, TCIOFLUSH)) {
		perror("tcflush");
		return -1;
	}
	if (tcsetattr(*port, TCSANOW, &term_st)) {
		perror("tcgetattr");
		return -1;
	}
	
	return 0;
}

void serial_port_flush(int fd)
{
	tcflush(fd, TCIOFLUSH);
}



src/amc.h
#ifndef _AMC_H_
#define _AMC_H_

#include <stdint.h>

#define AMC_SOF_BYTE 0xA5
#define AMC_CRC_POLY 0x1021

#define AMC_DEFAULT_TIMEOUT_MS 1000

#define AMC_EOK 0
#define AMC_ESERIALINIT -1
#define AMC_EINVALIDACCESSTYPE -2
#define AMC_EWRITE -3
#define AMC_EREAD -4
#define AMC_ETIMEOUT -5
#define AMC_ESEQ -6
#define AMC_ECRC -7
#define AMC_EINCOMPLETE -8
#define AMC_EINVALIDCMD -9
#define AMC_ENOACCESS -10
#define AMC_EFRAMEERR -11
#define AMC_EUNKNOWNSTATUS -12
#define AMC_EBUFSIZE -13

#define AMC_CMDTYPE_READ 1
#define AMC_CMDTYPE_WRITE 2
#define AMC_CMDTYPE_READWRITE 3

#define AMC_CMDRESP_COMPLETE 1
#define AMC_CMDRESP_INCOMPLETE 2
#define AMC_CMDRESP_INVALID 3
#define AMC_CMDRESP_NOACCESS 6
#define AMC_CMDRESP_FRAMEERR 8

/* TODO: make these macros aware of machine endianness */
#define amc_int16_to_le(x) x
#define amc_int16_from_le(x) x
#define amc_int32_to_le(x) x
#define amc_int32_from_le(x) x

/* Bridge control bits */
#define AMC_BC_INHIBIT (1 << 0)
#define AMC_BC_BRAKE (1 << 1)
#define AMC_BC_QUICKSTOP (1 << 6)
#define AMC_BC_RESETEVENTS (1 << 12)

/* Bridge status bits */
#define AMC_BS_ENABLED (1 << 0)
#define AMC_BS_DYNBRAKE (1 << 1)
#define AMC_BS_SHUNT (1 << 2)
#define AMC_BS_POSSTOP (1 << 3)
#define AMC_BS_NEGSTOP (1 << 4)
#define AMC_BS_POSTORQUEINH (1 << 5)
#define AMC_BS_NEGTORQUEINH (1 << 6)
#define AMC_BS_EXTBRAKE (1 << 7)

/* Drive protection status */
#define AMC_PS_RESET (1 << 0)
#define AMC_PS_INTERROR (1 << 1)
#define AMC_PS_SHORTCKT (1 << 2)
#define AMC_PS_OVERCURRENT (1 << 3)
#define AMC_PS_UNDERVOLTAGE (1 << 4)
#define AMC_PS_OVERVOLTAGE (1 << 5)
#define AMC_PS_OVERTEMP (1 << 6)

/* System protection status */
#define AMC_SS_RESTOREERR (1 << 0)
#define AMC_SS_STOREERR (1 << 1)
#define AMC_SS_MOTOROVERTEMP (1 << 4)
#define AMC_SS_FEEDBACKERROR (1 << 6)
#define AMC_SS_OVERSPEED (1 << 7)
#define AMC_SS_COMMERR (1 << 10)

/* Drive System status 1 */
#define AMC_DS_LOGMISSED (1 << 0)
#define AMC_DS_CMDINHIBIT (1 << 1)
#define AMC_DS_USERINHIBIT (1 << 2)
#define AMC_DS_POSINH (1 << 3)
#define AMC_DS_NEGINH (1 << 4)
#define AMC_DS_CURRENTLIM (1 << 5)
#define AMC_DS_CONTCURRENT (1 << 6)
#define AMC_DS_CLSAT (1 << 7)
#define AMC_DS_CMDDYNBRAKE (1 << 12)
#define AMC_DS_USERDYNBRAKE (1 << 13)
#define AMC_DS_SHUNTREG (1 << 14)

/* Drive System status 2 */
#define AMC_DS_ZEROVEL (1 << 0)
#define AMC_DS_ATCMD (1 << 1)
#define AMC_DS_VELOCITYERR (1 << 2)
#define AMC_DS_POSVELOCITYLIM (1 << 3)
#define AMC_DS_NEGVELOCITYLIM (1 << 4)
#define AMC_DS_CMDPROFILER (1 << 5)

struct amc_drive {
	uint8_t seq_ctr; /**< Message sequence counter */
	uint16_t *crc_table; /**< Cached CRC table */
	int device; /**< Device number for the communications port */
	int address; /**< Device address */
	int timeout_ms; /**< Read timeout in milliseconds */
	int debug; /**< Debug flag, set to 1 to enable debug messages */
};

union amc_control {
	uint8_t byte;
	struct {
		uint8_t cmd : 2; /**< Command type */
		uint8_t seq : 4; /**< Sequence number, rolls over at 0x0F */
		uint8_t rsvd : 2;
	} __attribute__((__packed__)) bits;
} __attribute__((__packed__));

/**
\brief Command packet sent to AMC drives

The destination address is set to 0x00 for broadcast, 0x01 - 0x3F are valid
destination addresses, and 0xFF is only valid for slave-to-master messages.

The command type is 0 = not used, 1 = message contains no data but the
response will contain the number of words specified in payload_len, 
2 = message contains number of words specified in payload_len but the response
will not contain any data, 3 = message and response both contain payloads of
length payload_len.
*/
struct amc_command {
	uint8_t sof; /**< Start of Frame (always set to 0xA5) */
	uint8_t addr; /**< Destination address, 00 = broadcast, 01-3F are valid, FF = Master */
	union amc_control control; /**< Control byte */
	uint8_t index; /**< Index into the parameter array within a drive */
	uint8_t offset; /**< Offset within the parameter array addressed by index */
	uint8_t payload_len; /**< Payload length in words */
	uint16_t crc; /**< CRC of the header */
} __attribute__((__packed__));

/**
\brief Response packet sent to AMC drives

*/
struct amc_response {
	uint8_t sof; /**< Start of Frame (always set to 0xA5) */
	uint8_t addr; /**< Destination address, 00 = broadcast, 01-3F are valid, FF = Master */
	union amc_control control; /**< Control byte */
	uint8_t status1; /**< First status byte */
	uint8_t status2; /**< Second status byte */
	uint8_t payload_len; /**< Payload length in words */
	uint16_t crc; /**< CRC of the header */
} __attribute__((__packed__));

struct amc_product_info {
	uint8_t rsvd1[2];
	uint8_t control_board_name[32];
	uint8_t control_board_version[32];
	uint8_t control_board_serial[32];
	uint8_t control_board_build_date[32];
	uint8_t control_board_build_time[32];
	uint8_t rsvd2[30];
	uint8_t product_part_number[32];
	uint8_t product_version[32];
	uint8_t product_serial_number[32];
	uint8_t product_build_date[32];
	uint8_t product_build_time[32];
} __attribute__((__packed__));

int amc_serial_open(char *dev, int spd);
int amc_drive_new(struct amc_drive *drv, int address, int serial_fd);
int amc_cmd_write(struct amc_drive *drv, struct amc_command *cmd, int access_type, 
	int response_len, uint16_t *payload, int payload_len);
int amc_resp_read(struct amc_drive *drv, struct amc_response *rsp, void *payload, int payload_max_size);

int amc_get_string(struct amc_drive *drv, int index, int offset, void *buffer, int bufsize);
int amc_get_uint16(struct amc_drive *drv, int index, int offset, uint16_t *buffer);
int amc_get_uint32(struct amc_drive *drv, int index, int offset, uint32_t *buffer);

int amc_write_string(struct amc_drive *drv, int index, int offset, void *buffer, int bufsize);
int amc_write_uint16(struct amc_drive *drv, int index, int offset, uint16_t value);
int amc_write_uint32(struct amc_drive *drv, int index, int offset, uint32_t value);

int amc_get_access_control(struct amc_drive *drv);
int amc_get_product_info(struct amc_drive *drv, struct amc_product_info *pi);
int amc_get_command_param(struct amc_drive *drv, unsigned int param, uint32_t *buffer);

#endif /* _AMC_H_ */

